### DISCUSSION TITLE
Django Output PDF without Saving to Server

### DISCUSSION THREAD

**Discussion by senzlord** [py-pdf/fpdf2#374] [Q&A]
I already following this https://python.plainenglish.io/generate-and-serve-pdf-files-with-django-e3efd9fde7bc
but there are some issue what if there are multiple user generate the PDF on same times the first user will get same like the second one if the filenames is same...

Is there anyway to make output not saved to server first but send to front end immediately?

Code I using now with saving to server first:
```
pdf.output('Quiz Report.pdf', 'F')
return FileResponse(open('Quiz Report.pdf', 'rb'), as_attachment=True, content_type='application/pdf')
```

Code I already try without saving to server:
```
pdf_as_bytes = pdf.output()
return FileResponse(pdf_as_bytes, as_attachment=True, content_type='application/pdf')
```

Result:
![image](https://user-images.githubusercontent.com/22251174/160382942-30dd406b-f14c-4c77-afc4-245f14ed148f.png)

Second Result: (If I remove content_type)
![image](https://user-images.githubusercontent.com/22251174/160383027-a3ef6f6a-864a-42ef-b154-4c81007e531c.png)

**Comment by Lucas-C** [comment]
Thank you for opening a new discussion as a follow-up of [#116](https://github.com/PyFPDF/fpdf2/issues/116)

As suggested previously, could you provide some minimal (but fully functional) Djanjo app reproducing your problem, please ?
_cf._ [stackoverflow.com/help/minimal-reproducible-example](https://stackoverflow.com/help/minimal-reproducible-example)

With only 2 lines of Python code provided, I won't be able to help much...

**Reply by senzlord** [reply]
This is the one I actually use, but only create the Header without my looping data.

```python
pdf = FPDF()
# Top
pdf.add_page(orientation="landscape", format="A4")
pdf.set_font('helvetica', size=12)
line_height = pdf.font_size * 2
col_width = pdf.epw / 10
pdf.cell(txt="Quiz Report")
pdf.ln(line_height)

# Header Table
pdf.set_font(style="B")
pdf.cell(col_width*1.5, line_height*2, 'Event', border=2, align='C')
pdf.cell(col_width*1.3, line_height*2, 'Agent', border=2, align='C')
pdf.cell(col_width, line_height*2, 'Total Tickets', border=2, align='C')
col = pdf.x - pdf.l_margin
pdf.cell(col_width/1.5*3, line_height, 'Total Winning Tickets', border=2, align='C')
pdf.cell(col_width, line_height*2, 'Total Bets', border=2, align='C')
pdf.cell(col_width*1.5, line_height*2, 'Total Prizes', border=2, align='C')
colfoot = pdf.x - pdf.l_margin
pdf.cell(col_width*1.5, line_height*2, 'P/L', border=2, align='C')
pdf.ln(line_height)
pdf.cell(col)
pdf.cell(col_width/1.5, line_height, 'Gold', border=2, align='C')
pdf.cell(col_width/1.5, line_height, 'Silver', border=2, align='C')
pdf.cell(col_width/1.5, line_height, 'Bronze', border=2, align='C')
pdf.ln(line_height)

# Footer Table
Total = 0
pdf.cell(colfoot, line_height, 'Grand Total', align='R')
pdf.set_text_color(220, 0, 0) if Total < 0 else pdf.set_text_color(0, 220, 0)
pdf.cell(col_width*1.5, line_height, f'{Total:,}', border=2, align='C')

pdf.output('Quiz Report.pdf', 'F')
return FileResponse(open('Quiz Report.pdf', 'rb'), as_attachment=True, content_type='application/pdf')
```

My Change only the last 2 lines like my post, if i remove content_type it will show the number

```
pdf_as_bytes = pdf.output()
return FileResponse(pdf_as_bytes, as_attachment=True, content_type='application/pdf')
```

**Comment by Lucas-C** [comment]
You did not provide a **fully functional** Django app 😔

Still, I was able to build one. There is the code you want to use:

```python
from fpdf import FPDF
from django.http import HttpResponse
from django.shortcuts import render

def report(request):
    ...
    return HttpResponse(bytes(pdf.output()), content_type='application/pdf')
```

**Comment by ryanshow** [comment]
Often times, this category of problem is solved by writing data to a BytesIO object instead of a file. This gives you a little bit more flexibility, as a BytesIO object generally behaves as a File object does, but its entirely stored in local memory. In this specific example, it doesn't matter too much because HttpResponse accepts a bytestring, so you can just use that method as @Lucas-C has described. If you want to use FileResponse, you would need to generate one of these BytesIO buffers:

```python
import io
from fpdf import FPDF
from django.http import FileResponse

def report(request):
    ...
    buf = io.BytesIO(pdf.output(dest='S'))
    buf.seek(io.SEEK_SET) # Moves the cursor back to the start of the file.
    return FileResponse(buf, is_attachment=True, filename='Quiz Report.pdf')
```

**Reply by Lucas-C** [reply]
Interesting!

@ryanshow: would you be interested in submitting a PR to improve the actual documentation by adding your snippet?
_cf._ https://pyfpdf.github.io/fpdf2/UsageInWebAPI.html

**Reply by ryanshow** [reply]
@Lucas-C I don't mind at all. I'll put something together later this week and submit a PR.
